src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs

```diff
@@ -5606,5 +5606,87 @@ static async Task Main()
             var comp = CSharpTestBase.CreateCompilation(source, options: TestOptions.ReleaseExe);
             CompileAndVerify(comp, expectedOutput: "StructAwaitable");
         }
+
+        [Fact, WorkItem(40251, "https://github.com/dotnet/roslyn/issues/40251")]
+        public void AssignRefAfterAwait()
+        {
+            const string source = @"
+using System.Threading.Tasks;
+using System;
+
+class IntCode
+{
+    public static async Task Main()
+    {
+        await Step(0);
+    }
+
+    public static async Task CompletedTask()
+    {
+    }
+
+    public static async Task Step(int i)
+    {
+        Console.Write(field);
+        await CompletedTask();
+        ReadMemory() = i switch
+        {
+            _ => GetValue()
+        };
+        Console.Write(field);
+    }
+
+    public static long GetValue()
+    {
+        Console.Write(2);
+        return 3L;
+    }
+
+    private static long field;
+    private static ref long ReadMemory()
+    {
+        Console.Write(1);
+        return ref field;
+    }
+}
+";
+            var diags = new[]
+            {
+                // (12,30): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
+                //     public static async Task CompletedTask()
+                Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "CompletedTask").WithLocation(12, 30)
+            };
+
+            CompileAndVerify(source, options: TestOptions.DebugExe, verify: Verification.Skipped, expectedOutput: "0123").VerifyDiagnostics(diags);
+            CompileAndVerify(source, options: TestOptions.ReleaseExe, verify: Verification.Skipped, expectedOutput: "0123").VerifyDiagnostics(diags);
+        }
+
+        [Fact, WorkItem(40251, "https://github.com/dotnet/roslyn/issues/40251")]
+        public void AssignRefWithAwait()
+        {
+            const string source = @"
+using System.Threading.Tasks;
+
+class IntCode
+{
+    public async Task Step(Task<int> t)
+    {
+        ReadMemory() = await t;
```

Comment:
Consider adding a test where we have a `ref` l-value and use a compound operator. Essentially 

```cs
ReadMemory() += await t;
```